### ISSUE TITLE
Enable Facebook oembeds

### ISSUE THREAD

**Issue by gabu69** [Automattic/facebook-instant-articles-wp#556]
Wordpress 4.7 has just been released and it supports oEmbed for Facebook posts/activities/photo/videos/media/questions/notes. 

However while using this new feature, Facebook **instant articles plugin won't show up the embed** in the instant articles. 

Please help to integrate them to Instant articles since we are forced to use for example the videos that are reuploaded to Youtube, Vimeo etc

**Comment by blackguyinablackroom** [comment]
hi, i'm having a problem regarding jetpack and fb instant article plugins
i'm trying to post an fb public post to a wp blog site as a reference for an article but the problem is this message saying "No rules defined for in the context of InstantArticle" whenever i updated it.
i tried to use , it works but my boss doesn't like that method since i can post a twitter post with just the use of URL. Now he wants me to embed it (fb post) like i did on twitter post. As for the jetpack, i installed, activated and connected it to wordpress.com as per the instructions here https://optimizerwp.com/wordpress-embed-guide/, but its always the FB IA that says "No rules defined for in the context of InstantArticle". I've been trying to solve this problem for weeks now and i'm already tired in searching for answers. You guys are my last hope in resolving this problem.

thanks,

**Comment by abdusfauzi** [comment]
@gabu69 i just found out about this when one of my editors mentioned that FB Video embed didn't show up in Instant Articles.

Using this filter for WordPress, I managed to get it going and added into Instant Articles:

```
function social_embed_type_additional( $provider_name, $url ) {
    if ( ! class_exists( 'WP_oEmbed' ) ) {
        include_once( ABSPATH . WPINC . '/class-oembed.php' );
    }

    // Instead of checking all possible URL variants, use the provider list from WP_oEmbed.
    $wp_oembed = new WP_oEmbed();
    $provider_url = $wp_oembed->get_provider( $url );

    if ( false !== strpos( $provider_url, 'facebook.com' ) ) {
        $provider_name = 'facebook';
    }

    return $provider_name;
}
add_filter( 'instant_articles_social_embed_type', 'social_embed_type_additional', 10, 2 );
```

**Comment by blackguyinablackroom** [comment]
@gabu69 in what file did u put this code?? thanks xD

**Comment by abdusfauzi** [comment]
@blackguyinablackroom just paste this code in the `functions.php` file, which should be available in you theme folder. 😄

**Comment by blackguyinablackroom** [comment]
@gabu69 how did u embed the fb video?? did u use the [embed][/embed] shortcode??

**Comment by abdusfauzi** [comment]
@blackguyinablackroom by adding support for url detection, WordPress will convert any known URL structure into oEmbed codes (ex: YouTube iframe).

Read more https://codex.wordpress.org/Embeds

**Comment by blackguyinablackroom** [comment]
But the problem is the dB instant article

**Comment by blackguyinablackroom** [comment]
@gabu69 yes, it works (fb instant article deactivated).... but the problem is the fb instant article, this error always appear "No rules defined for <post> in the context of Paragraph" whenever i try to use url

**Comment by abdusfauzi** [comment]
@blackguyinablackroom that's weird. My issue was solved immediately with the code I added (above). Can you share the screenshot/snippet of the warning by Instant Articles?

And one more thing, you perhaps should tag @abdusfauzi to reply me and not @gabu69 😅

**Comment by blackguyinablackroom** [comment]
@abdusfauzi sorry dude haha.... by the way this is the snippet of the error i'm talking about 
![image](https://cloud.githubusercontent.com/assets/24713233/22538644/ee0af988-e94e-11e6-88d9-d220fa47d5ae.png)

**Comment by abdusfauzi** [comment]
@blackguyinablackroom i used to get that too. Try to update to the latest version  👌🏻 They merged the fix already.

**Comment by abdusfauzi** [comment]
@blackguyinablackroom i think i know why. It was a typo error on the variable `$providerURL` instead of `$provider_URL`.

On line 42, 44 and 46.
https://github.com/Automattic/facebook-instant-articles-wp/blob/master/embeds.php#L42

Sorry 😅

**Comment by blackguyinablackroom** [comment]
still not working for me T.T

**Comment by abdusfauzi** [comment]
@blackguyinablackroom you manually fix the typo, and it is not working? :(

**Comment by blackguyinablackroom** [comment]
yeah....... dunno why... i changed the providerURL to provider_url in line
42

**Comment by abdusfauzi** [comment]
@blackguyinablackroom you need to fix 3 lines: 42, 44, 46.

code that effectively detecting Facebook url is at line 46.

**Comment by blackguyinablackroom** [comment]
@abdusfauzi dude... still not working for me

**Comment by abdusfauzi** [comment]
@blackguyinablackroom email me at abdusfauzi@exnano.io. Lets try work on this exclusively on your WordPress setup. Create a new account 'abdusfauzi' with access to the articles. :)

**Comment by gabu69** [comment]
Ok I dediced to give it a try at 3.3.3 and its working, only con is its not full width.

So I guess not its to make it full width just like any other video embed.

**Comment by abdusfauzi** [comment]
@gabu69 same with me. All other video embed (Vimeo, YouTube) manage to get full-width in Instant Articles. I guess, Facebook (app) rendering is the issue.

**Comment by arnimation** [comment]
hi @abdusfauzi hope you can also help me with this error:

> No rules defined for <div class="fb-video"> in the context of InstantArticle

I have Instant Articles for WP plugin installed and updated 4.0.2 version.  I installed through wordpress plugins page.

we basically just paste the fb video url and wordpress converts to something like below:

<p>Check out the video below to see what I mean</p>
<div class="fb-video" data-allowfullscreen="true" data-href="https://www.facebook.com/..../videos/..../"></div>

Then we get the error - No rules defined for <div class="fb-video"> in the context of InstantArticle

Not sure how to fix it. Thanks in advance!

**Comment by abdusfauzi** [comment]
@arnimation that's weird. I tested locally and production, and all is good. Do you mind share your FB Video url?

**Comment by arnimation** [comment]
@abdusfauzi sure, you can view it on this blog: https://pulutan.club/videos/calling-all-the-seafood-lovers/

Then it was shared here: https://www.facebook.com/pulutan.club/ already converted to FB IA but embedded facebook video did not transform.

On FB Add, you can see the FB IA https://www.facebook.com/pulutan.club/posts/922801317860766

Thanks again.

**Comment by abdusfauzi** [comment]
@arnimation hmm. thats weird. The video turn out fine on my WP setup. Is there any other plugin that might interfere with video embed or facebook embed? And I found that there's supposed to be a `<script></script>` element just before the `<div class"fb-video...` code. But it was missing in your post.

Unlike others, Facebook oEmbed codes are not `<iframe></iframe>` but rather some javascript with `<blockquote>`. You can see the HTML here:

https://www.facebook.com/plugins/video/oembed.json/?url=https://www.facebook.com/BestEverFoodReviewShow/videos/1325227197594040/

**Comment by arnimation** [comment]
@abdusfauzi thanks for the patience. its really weird.  maybe the following details will help:

This is a wordpress.com business plan site (recently opened up to supporting plugins) https://wordpress.com/pricing/

Plugins:

- Advanced Ads

- AMP

- Instant Articles for WP

- Jetpack

- Vaultpress

- Wordpress Importer

That's it.  It's a relatively new installation.  Probably a few weeks old.

**Comment by abdusfauzi** [comment]
@arnimation i'm taking a blind guess here, can you see if Jetpack has any "embed" feature enabled?can you disable them?

or try disabled each one of them at a time.

**Comment by arnimation** [comment]
@abdusfauzi thanks, will try that out and share if it fixes any. thanks again.

**Comment by Gelurban** [comment]
Hello @arnimation Did you manage to solve the problem?
I'm having the same issue:
No rules defined for <div class="fb-video"> in the context of InstantArticle

**Comment by arnimation** [comment]
Hi @abdusfauzi - disabled all my plugins and still no luck :( Checked jetpack and there's no embed feature.  Not sure what else to do. :(

Hi @Gelurban - no luck

**Comment by arnimation** [comment]
hi @abdusfauzi if we just add a transformer rule for the fb-video issue? what will it be? thanks in advance.

**Comment by mautematico** [comment]
Yes, Jetpack has embed feature that can't be disabled on WordPress 4.8; disabling Jetpack, fb-video issue gets solved on this plugin.


this an example source input fb-ia gets when Jetpack is disabled:
`<div class="embed">
    <div id="fb-root"></div>
    <p>
        <script>
            (function (d, s, id) { var js, fjs = d.getElementsByTagName(s)[0]; if (d.getElementById(id)) return; js = d.createElement(s); js.id = id; js.src = "//connect.facebook.net/en_US/sdk.js#xfbml=1&version=v2.10"; fjs.parentNode.insertBefore(js, fjs); }(document, 'script', 'facebook-jssdk'));
        </script>
    </p>
    <div class="fb-video" data-href="https://www.facebook.com/xxxxxx/videos/xxxxxx" data-width="688">
        <blockquote cite="https://www.facebook.com/xxxxxx/videos/xxxxxx/" class="fb-xfbml-parse-ignore">
            <p><a href="https://www.facebook.com/xxxxxx/videos/xxxxxx/"></a></p>
            <p>html escaped video description</p>
            <p>Posted by Facebook User on Thursday, July 27, 2017</p>
        </blockquote>
    </div>
</div>`

which gets properly transformed into an IA video.

However, this is the source input fb-ia gets when Jetpack is enabled for that same video
`<div class="fb-video" data-allowfullscreen="true" data-href="https://www.facebook.com/xxxxx/videos/xxxxxx"></div>`

which won't be transformed at all, but supressed.

**Comment by Gelurban** [comment]
@mautematico And is there any way of fixing this problem without disabling Jetpack?
Maybe a custom transformer rule?

**Comment by mautematico** [comment]
Sure it can be fixed without disabling Jetpack; I still did not find the propper way to do it.
